README.md
# Mimecon
A package to bind an icon to a mimeType

## Pull requests
Pull requests are welcome, I'm a beginner so don't hesitate if you see any corrections

## Usage

- Insert a mimetype label (eg: "image/png", "application/pdf"...)
- Defined some parameters a size, color, isOutlined etc if you want 
- An icon who match with the mimetype label will be return 


## Example

```dart
 Mimecon(
    mimetype: "application/png",
    color: Colors.red,
    size: 25,
    isOutlined: true,
  ),
```
## Additional information

- All icons come from material_design_icons_flutter package which themselves come from https://materialdesignicons.com/

- https://github.com/romgrm/mimecon_package

- for example project pls visit [mimecon_example](https://github.com/romgrm/mimecon_example)

CHANGELOG.md
# Changelog

## [1.0.1] - 07/08/2023
* Upgraded dependencies | put mimeconType nullable | cleaning some stuffs
## [0.0.2] - 27/05/2022
* Updated pubspec + format for pass analytics 
## [0.0.1] - 27/05/2022
* Initial release 

test/mimecon_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';

import 'package:mimecon/mimecon.dart';

void main() {
  testWidgets('Mimecon should return default icon', (tester) async {
    await tester.pumpWidget(const MaterialApp(
      home: Scaffold(
          body: Mimecon(
        mimetype: "image/png",
      )),
    ));

    final iconFinder = find.byIcon(MdiIcons.image);

    expect(iconFinder, findsOneWidget);
    expect((tester.firstWidget(find.byType(Icon)) as Icon).color, const Color(0xFF858593));
    expect((tester.firstWidget(find.byType(Icon)) as Icon).size, 40);
  });

  testWidgets('Mimecon should return custom icon', (tester) async {
    await tester.pumpWidget(const MaterialApp(
      home: Scaffold(
          body: Mimecon(
        mimetype: "image/png",
        color: Colors.red,
        size: 20,
      )),
    ));

    expect((tester.firstWidget(find.byType(Icon)) as Icon).color, Colors.red);
    expect((tester.firstWidget(find.byType(Icon)) as Icon).size, 20);
  });
  testWidgets('Mimecon should return outlined icon', (tester) async {
    await tester.pumpWidget(const MaterialApp(
      home: Scaffold(
          body: Mimecon(
        mimetype: "image/png",
        isOutlined: true,
      )),
    ));

    final iconFinder = find.byIcon(MdiIcons.imageOutline);

    expect(iconFinder, findsOneWidget);
  });
  testWidgets('Mimecon should return generic icon when mimetype is null', (tester) async {
    await tester.pumpWidget(const MaterialApp(
      home: Scaffold(body: Mimecon(mimetype: null)),
    ));

    final iconFinder = find.byIcon(MdiIcons.file);

    expect(iconFinder, findsOneWidget);
  });
}


lib/mimecon.dart
library mimecon;

import 'dart:developer' as devlog;

import 'package:flutter/material.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import 'package:mimecon/mimetype.dart';

/// Widget that return icon from a specific mimeType.
///
/// All icons come from material_design_icons_flutter package which themselves come from https://materialdesignicons.com/
class Mimecon extends StatefulWidget {
  /// The specific mimeType (eg: image/png, video/mp4, application/pdf...)
  ///
  /// See mimetype.dart for all mimetypes listed
  final String? mimetype;

  /// Color of icon (only with Color widget)
  final Color color;

  /// Size of icon
  final double size;

  /// Defined if the icon is outlined or filled
  final bool isOutlined;

  const Mimecon({Key? key, this.mimetype, this.color = Colors.black54, this.size = 40.0, this.isOutlined = false}) : super(key: key);

  @override
  State<Mimecon> createState() => _MimeconState();
}

class _MimeconState extends State<Mimecon> {
  @override
  void initState() {
    super.initState();
    catchErrorMimetype(widget.mimetype);
  }

  @override
  Widget build(BuildContext context) {
    return _setIcon(widget.mimetype, widget.color, widget.size, widget.isOutlined);
  }

  void catchErrorMimetype(String? mimetype) {
    if (mimetype == null) {
      devlog.log("Default icon output return");
    } else {
      RegExp mimetypePattern = RegExp(r'^[-\w.]+/[-\w.]+$');
      if (!mimetypePattern.hasMatch(mimetype)) {
        devlog.log("The entered mimetype does not correspond to a standard mimetype format");
        devlog.log("Default icon output return");
      }
    }
  }

  Icon _setIcon(String? mimeType, color, size, isOutlined) {
    try {
      if (imageMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.imageOutline : MdiIcons.image,
          color: color,
          size: size,
        );
      } else if (videoMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.filmstripBox,
          color: color,
          size: size,
        );
      } else if (audioMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.musicBoxOutline : MdiIcons.musicBox,
          color: color,
          size: size,
        );
      } else if (docMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.microsoftWord,
          color: color,
          size: size,
        );
      } else if (excelMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.microsoftExcel,
          color: color,
          size: size,
        );
      } else if (powerpointMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.microsoftPowerpoint,
          color: color,
          size: size,
        );
      } else if (googleDriveMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.googleDrive,
          color: color,
          size: size,
        );
      } else if (pdfMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.filePdfBox,
          color: color,
          size: size,
        );
      } else if (libreOfficeWriterMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.fileDocumentOutline : MdiIcons.fileDocument,
          color: color,
          size: size,
        );
      } else if (libreOfficeImpressMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.presentation,
          color: color,
          size: size,
        );
      } else if (libreOfficeCalcMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.fileTableOutline : MdiIcons.fileTable,
          color: color,
          size: size,
        );
      } else if (compressedFileMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.folderZipOutline : MdiIcons.folderZip,
          color: color,
          size: size,
        );
      } else if (ebookMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.bookshelf,
          color: color,
          size: size,
        );
      } else if (codeMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.fileCodeOutline : MdiIcons.fileCode,
          color: color,
          size: size,
        );
      } else if (calendarMimeType.contains(mimeType)) {
        return Icon(
          isOutlined ? MdiIcons.calendarOutline : MdiIcons.calendar,
          color: color,
          size: size,
        );
      } else if (jsonMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.codeJson,
          color: color,
          size: size,
        );
      } else if (fontMimeType.contains(mimeType)) {
        return Icon(
          MdiIcons.formatLetterCase,
          color: color,
          size: size,
        );
      }
    } catch (errorCatch) {
      devlog.log("Error in compilation", error: errorCatch);
    }

    return Icon(
      isOutlined ? MdiIcons.fileOutline : MdiIcons.file,
      color: color,
      size: size,
    );
  }
}


lib/mimetype.dart
List<String> imageMimeType = [
  "image/gif",
  "image/jpg",
  "image/jpeg",
  "image/png",
  "image/x-png",
  "image/tiff",
  "image/bmp",
  "image/svg+xml",
  "image/webp",
  "image/vnd.microsoft.icon",
  "image/webp",
];

List<String> videoMimeType = [
  "application/x-flash-video",
  "application/x-shockwave-flash",
  "video/mp4",
  "video/webm",
  "video/x-ms-asf",
  "video/x-msvideo",
  "video/x-matroska",
  "video/quicktime",
  "video/mpeg",
  "video/x-ms-wmv",
  "video/x-msvideo",
  "video/ogg",
  "video/mp2t",
  "video/3gpp2"
];

List<String> audioMimeType = [
  "audio/mp3",
  "audio/mpeg",
  "audio/mpeg3",
  "audio/x-mpeg-3",
  "audio/ogg",
  "audio/x-vorbis+ogg",
  "audio/mp4",
  "audio/wav",
  "audio/x-wav",
  "audio/x-ms-wma",
  "audio/aac",
  "audio/opus",
  "audio/webm",
  "audio/3gpp2"
];

List<String> docMimeType = [
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "application/msword",
  "application/vnd.ms-word",
  "application/doc",
  "application/text",
  "application/vnd.msword",
  "application/winword",
  "application/word",
  "application/x-msw6",
  "application/x-msword",
  "application/x-abiword",
  "application/x-freearc"
];

List<String> excelMimeType = [
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  "application/vnd.ms-excel",
  "application/msexcel",
  "application/x-msexcel",
  "application/x-ms-excel",
  "application/x-excel",
  "application/x-dos_ms_excel",
  "application/xls",
];

List<String> powerpointMimeType = [
  "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  "application/vnd.ms-powerpoint",
  "application/mspowerpoint",
  "application/ms-powerpoint",
  "application/mspowerpnt",
  "application/vnd-mspowerpoint",
  "application/powerpoint",
  "application/x-powerpoint",
  "application/x-m",
];

List<String> googleDriveMimeType = [
  "application/vnd.google-apps.document",
  "application/vnd.google-apps.spreadsheet",
  "application/vnd.google-apps.presentation",
];

List<String> pdfMimeType = [
  "application/pdf",
  "application/x-pdf",
  "application/acrobat",
  "application/vnd.pdf",
  "text/pdf",
  "text/x-pdf",
];

List<String> libreOfficeWriterMimeType = [
  "application/vnd.sun.xml.writer",
  "application/vnd.oasis.opendocument.text",
  "application/application/x-vnd.oasis.opendocument.text",
];

List<String> libreOfficeImpressMimeType = [
  "application/vnd.sun.xml.impress",
  "application/vnd.oasis.opendocument.presentation",
  "application/x-vnd.oasis.opendocument.presentation",
];

List<String> libreOfficeCalcMimeType = [
  "application/vnd.sun.xml.calc",
  "application/vnd.oasis.opendocument.spreadsheet",
  "application/x-vnd.oasis.opendocument.spreadsheet",
];

List<String> ebookMimeType = [
  "application/vnd.amazon.ebook",
  "application/epub+zip"
];

List<String> compressedFileMimeType = [
  "application/gzip",
  "application/zip",
  "application/x-bzip",
  "application/x-bzip2",
  "application/x-zip-compressed",
  "application/x-tar",
  "application/x-tar-gz",
  "application/vnd.rar",
  "application/x-7z-compressed",
];

List<String> codeMimeType = [
  "text/html",
  "text/css",
  "text/javascript",
  "application/java-archive",
  "application/x-httpd-php",
  "application/x-sh",
  "application/xml",
  "application/xhtml+xml",
];

List<String> calendarMimeType = ["text/calendar"];

List<String> jsonMimeType = [
  "application/json",
  "application/ld+json",
];

List<String> fontMimeType = [
  "font/otf",
  "font/ttf",
  "font/woff",
  "font/woff2",
];

List<String> genericMimeType = [
  "text/plain",
  "application/octet-stream",
  "text/csv",
  "text/vtt",
  "application/msword",
  "application/mswrite",
  "application/vnd.ms-fontobject",
  "application/ogg",
  "application/vnd.apple.installer+xml",
  "application/rtf",
  "application/x-shockwave-flash",
  "application/vnd.visio",
];


example/lib/main.dart
import 'package:example/entity/mimetypeEntity.dart';
import 'package:flutter/material.dart';
import 'package:mimecon/mimecon.dart';

void main() {
  runApp(const MimeconExample());
}

class MimeconExample extends StatefulWidget {
  const MimeconExample({Key? key}) : super(key: key);

  @override
  State<MimeconExample> createState() => _MimeconExampleState();
}

class _MimeconExampleState extends State<MimeconExample> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: SafeArea(
          child: Column(
            children: [
              Container(
                padding: const EdgeInsets.all(25),
                child: const Text("Mimecon Example"),
              ),
              Expanded(
                child: ListView.separated(
                  shrinkWrap: true,
                  itemCount: mimeTypeData.length,
                  itemBuilder: (context, index) {
                    String key = mimeTypeData.keys.elementAt(index);
                    return ListTile(
                      leading: Mimecon(
                        mimetype: mimeTypeData[key]!.mimetype,
                        color: mimeTypeData[key]!.color,
                        size: mimeTypeData[key]!.size,
                        isOutlined: mimeTypeData[key]!.isOutlined,
                      ),
                      title: Text(key),
                    );
                  },
                  separatorBuilder: (context, index) => const Divider(),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

Map<String, MimetypeEntity> mimeTypeData = {
  "Image jpg mimetype": MimetypeEntity(
      mimetype: "image/jpg", color: Colors.green, size: 40, isOutlined: false),
  "Pdf mimetype little":
      MimetypeEntity(mimetype: "application/pdf", color: Colors.blue, size: 25),
  "Pdf mimetype with only mimeType (default config)":
      MimetypeEntity(mimetype: "application/pdf"),
  "Html mimetype outlined ": MimetypeEntity(
      mimetype: "text/html", color: Colors.purple, isOutlined: true),
  "Generic mimetype ": MimetypeEntity(
      mimetype: "text/plain", color: Colors.brown, isOutlined: true),
};
